fix: add support for subpath exports

acandoo opened the pull request:
<!-- Oh wow! Thanks for opening a pull request! 😁 🎉 -->
<!-- You are very welcome here and any contribution is appreciated. 👍 -->
<!-- Choose one of the checklists if it applies to you and delete the rest. -->

## I'm fixing a bug or typo

- [x] if this is your first time contributing, run `npm run add-contributor` and follow the prompts to add yourself to the README
- [x] squash merge the PR with commit message "fix: [Description of fix]"

This PR allows for native ESM imports of `isomorphic-git/http/node` and `isomorphic-git/http/web`. This is especially useful for Node projects using ES modules, as they no longer have to specify `isomorphic-git/http/node/index.js` or use a bundler that can resolve the import.

Fixes #2115

jcubic commented:
The tests are passing, there were some failing tests, because it seems they are not stable.

Let me know when this is ready.

acandoo commented:
Okay, so doing this properly with types ended up being a more substantial change. I added `.d.cts` files to what typescript has to build and added explicit type paths to `package.json` for good measure. However, I can't make the main export an ES module until #2117 is resolved.

Unfortunately, I can't run the test suite without running out of memory on my computer, but based on a testing project that runs basic CJS/ESM and Node/browser test scripts, I don't think this breaks anything.

Let me know if I should make any changes!

jcubic commented:
Thanks for your contribution. I need to test this locally. I have never seen cts file before, I thought that you only allow using import with TypeScript.

jcubic commented:
Can you fix resolve conflicts? You need to remove your username from contributors and add it again.

acandoo commented:
okay rebased!

jcubic commented:
Sorry, I didn't have the time to test this. Will try to find some free time till the end of the week.

jcubic commented:
TypeScript + ESM doesn't work. Got this error:


> Error [ERR_UNSUPPORTED_DIR_IMPORT]: Directory import '/home/kuba/projects/jcubic/isomorphic-git/import/node_modules/isomorphic-git/http/node' is not supported resolving ES modules imported from /home/kuba/projects/jcubic/isomorphic-git/import/index.ts
Did you mean to import "isomorphic-git/http/node/index.cjs"?

I used the latest Node with optional typescript support.

acandoo commented:
Still seems fine to me... I've only tested on Linux so maybe it's a platform issue? Either way I don't think it would be an issue to solve on isomorphic-git's end.
```
// Node 22.17.0

$ node --experimental-strip-types index.ts
(node:85038) ExperimentalWarning: Type Stripping is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
0.0.0-development
```

```
// Node 24.4.0

$ node --experimental-strip-types index.ts
0.0.0-development
```

jcubic commented:
Can give exact steps how you run your project?

I use your branch, built it extracted the tar ball and use npm link to install in new project.

Did you use the same?

acandoo commented:
Here is the test repo I used: <https://github.com/acandoo/isomorphic-git-testing/>

Build isomorphic-git on my branch, use `npm link` in the repo, go into the testing repo, `npm install`. That works on my end.

acandoo commented:
Oh, never mind, I was able to replicate your error by using the tarball.

acandoo commented:
I tried it with the extracted tarball and it still worked.

jcubic commented:
Checking again, and it does work when using git repo for `npm link`.

jcubic commented:
The release failed. I'm not sure why. There is an error in clone, I've restarted the build. The failure is not related to your changes.
